File: courses/codes/cclv1-jr-class/ccc-lv1-sr-codes/first_non_repeating.py
from queue import Queue

def firstNonRepeating(s: str) -> None:
    # Initialize queue and frequency map
    queue = Queue()
    frequency_map = {}
    
    for c in s:
        if c not in frequency_map:  # Character not in frequency map
            frequency_map[c] = 0
        frequency_map[c] += 1
        
        if frequency_map[c] == 1:  # c is unique so far
            queue.put(c)
        
        # Remove repeating characters from the front of the queue
        while not queue.empty() and frequency_map[queue.queue[0]] > 1:
            queue.get()
        
        # Print the first non-repeating character or 0 if none
        if queue.empty():
            print('0', end="")
        else:
            print(queue.queue[0], end="")

# Example usage
s = "aabc"
firstNonRepeating(s)  # Output: a00b

File: courses/codes/cclv1-jr-class/ccc-lv1-sr-codes/FirstNonRepeating.java
import java.util.*;

public class FirstNonRepeating {
    public static void firstNonRepeating(String s) {
        // Initialize queue and frequency map
        Queue<Character> queue = new LinkedList<>();
        Map<Character, Integer> frequencyMap = new HashMap<>();
        
        for (char c : s.toCharArray()) {
            frequencyMap.put(c, frequencyMap.getOrDefault(c, 0) + 1);
            
            if (frequencyMap.get(c) == 1) {  // c is unique so far
                queue.add(c);
            }
            
            // Remove repeating characters from the front of the queue
            while (!queue.isEmpty() && frequencyMap.get(queue.peek()) > 1) {
                queue.poll();
            }
            
            // Print the first non-repeating character or 0 if none
            if (queue.isEmpty()) {
                System.out.print("0");
            } else {
                System.out.print(queue.peek());
            }
        }
    }

    public static void main(String[] args) {
        String s = "aabc";
        firstNonRepeating(s);  // Output: a00b
    }
}
